test/e2e/node/pod_resize.go

```diff
@@ -0,0 +1,1421 @@
+						{"name":"c3", "resources":{"requests":{"cpu":"100m","memory":"100Mi"},"limits":{"cpu":"200m","memory":"200Mi"}}}
+					]}}`,
+			expected: []TestContainerInfo{
+				{
+					Name:      "c1",
+					Resources: &ContainerResources{CPUReq: "100m", CPULim: "200m", MemReq: "100Mi", MemLim: "200Mi"},
+					CPUPolicy: &doRestart,
+					MemPolicy: &doRestart,
+				},
+				{
+					Name:         "c2",
+					Resources:    &ContainerResources{CPUReq: "250m", CPULim: "350m", MemReq: "250Mi", MemLim: "350Mi"},
+					CPUPolicy:    &noRestart,
+					MemPolicy:    &noRestart,
+					RestartCount: 1,
+				},
+				{
+					Name:         "c3",
+					Resources:    &ContainerResources{CPUReq: "100m", CPULim: "200m", MemReq: "100Mi", MemLim: "200Mi"},
+					CPUPolicy:    &doRestart,
+					MemPolicy:    &doRestart,
+					RestartCount: 1,
+				},
+			},
+		},
+	}
+
+	for idx := range tests {
+		tc := tests[idx]
+		ginkgo.It(tc.name, func() {
+			var testPod, patchedPod *v1.Pod
+			var pErr error
+
+			tStamp := strconv.Itoa(time.Now().Nanosecond())
+			initDefaultResizePolicy(tc.containers)
+			initDefaultResizePolicy(tc.expected)
+			testPod = makeTestPod(ns, "testpod", tStamp, tc.containers)
+
+			ginkgo.By("creating pod")
+			newPod := podClient.CreateSync(testPod)
+
+			ginkgo.By("verifying the pod is in kubernetes")
+			selector := labels.SelectorFromSet(labels.Set(map[string]string{"time": tStamp}))
+			options := metav1.ListOptions{LabelSelector: selector.String()}
+			podList, err := podClient.List(context.TODO(), options)
+			framework.ExpectNoError(err, "failed to query for pods")
+			gomega.Expect(len(podList.Items) == 1)
+
+			ginkgo.By("verifying initial pod resources, allocations, and policy are as expected")
+			verifyPodResources(newPod, tc.containers)
+			verifyPodResizePolicy(newPod, tc.containers)
+
+			ginkgo.By("verifying initial pod status resources and cgroup config are as expected")
+			verifyPodStatusResources(newPod, tc.containers)
+			verifyPodContainersCgroupValues(newPod, tc.containers, true)
+
+			ginkgo.By("patching pod for resize")
+			patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(context.TODO(), newPod.Name,
+				types.StrategicMergePatchType, []byte(tc.patchString), metav1.PatchOptions{})
+			framework.ExpectNoError(pErr, "failed to patch pod for resize")
+
+			ginkgo.By("verifying pod patched for resize")
+			verifyPodResources(patchedPod, tc.expected)
+			verifyPodAllocations(patchedPod, tc.containers, true)
+
+			ginkgo.By("waiting for resize to be actuated")
+			resizedPod := waitForPodResizeActuation(newPod, patchedPod, tc.expected)
+
+			ginkgo.By("verifying pod container's cgroup values after resize")
+			verifyPodContainersCgroupValues(resizedPod, tc.expected, true)
+
+			ginkgo.By("verifying pod resources after resize")
+			verifyPodResources(resizedPod, tc.expected)
+
+			ginkgo.By("verifying pod allocations after resize")
+			verifyPodAllocations(resizedPod, tc.expected, true)
+
```

Comment:
testCase I added a field: resizeToOriginalPatchString;
The test case is that resource A to resource B resize is successful, and then resize back to resource A. The code is as follows


if tc.resizeToOriginalPatchString != "" {
		ginkgo.By("patching pod for resize Original resource")
		patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(context.TODO(), newPod.Name,
			types.StrategicMergePatchType, []byte(tc.resizeToOriginalPatchString), metav1.PatchOptions{})
		framework.ExpectNoError(pErr, "failed to patch pod for resize Original resource")
		ginkgo.By("verifying pod patched for resize Original resource")
		verifyPodResources(patchedPod, tc.containers)
		verifyPodAllocations(patchedPod, tc.expected, true)
		ginkgo.By("waiting for resize to be actuated Original resource")
		resizedPod := waitForPodResizeActuation(newPod, patchedPod, tc.containers)
		ginkgo.By("verifying pod container's cgroup values after resize Original resource")
		verifyPodContainersCgroupValues(resizedPod, tc.containers, true)
		ginkgo.By("verifying pod resources after resize Original resource")
		verifyPodResources(resizedPod, tc.containers)
		ginkgo.By("verifying pod allocations after resize Original resource")
		verifyPodAllocations(resizedPod, tc.containers, true)
}


**I tested a case and the results are as follows**


• Failure [254.695 seconds]
[sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling]
/home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/framework.go:23
  Guaranteed QoS pod, one container - increase CPU & memory [It]
  /home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/pod_resize.go:1151

  May  5 16:09:39.017: failed to verify container cgroup values equals expected values
  Unexpected error:
      <*errors.errorString | 0xc0000d0180>: {
          s: "timed out waiting for container cgroup values to match expected",
      }
      timed out waiting for container cgroup values to match expected
  occurred

  /home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/pod_resize.go:412
------------------------------
{"msg":"FAILED [sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] Guaranteed QoS pod, one container - increase CPU \u0026 memory","total":1,"completed":0,"skipped":0,"failed":1,"failures":["[sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] Guaranteed QoS pod, one container - increase CPU \u0026 memory"]}
May  5 16:09:39.248: INFO: Running AfterSuite actions on all nodes
May  5 16:09:39.249: INFO: Running AfterSuite actions on node 1
{"msg":"Test Suite completed","total":1,"completed":0,"skipped":0,"failed":1,"failures":["[sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] Guaranteed QoS pod, one container - increase CPU \u0026 memory"]}


Summarizing 1 Failure:

[Fail] [sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] [It] Guaranteed QoS pod, one container - increase CPU & memory 
/home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/pod_resize.go:412

Ran 1 of 1 Specs in 254.848 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestE2E (254.85s)
FAIL

Reply:
Thanks very much for this test case! I think this should be added to the E2E test. I'll try it out and take a closer look at it. But upon reading the test, I suspect this is caused not by the hash, but rather by a condition check in computePodResizeAction where we don't generate a resize action if spec.Resources == status.Resources.
```golang
blablabla...
    if ..
    ..
    len(diff.ObjectDiff(container.Resources, apiContainerStatus.Resources)) == 0
    ...
    return true
blablabla
```
Currently, containerd does not implement ContainerStatus CRI extension needed to have status.Resources reflect the cgroup values on the running container. @ruiwen-zhao has a draft PR ready for this but it can't go in without this PR. And this test cannot be activated without his change... a bit of a chicken-egg problem here.

But I think this will go away once Ruiwen's containerd change is merged. (Again, I'll take a closer look to verify this is indeed the issue). 

Regardless, I think your test case is very valuable, thank you! 
Would you be able to create a PR for this test case (after my PR is merged)? (Alternately, you can send a PR to my PR and I can credit you as collaborator .. but I prefer tracking it and doing it as a separate change as this is not alpha blocker).

Reply:
I've created a beta-blocker issue to track adding this test case. https://github.com/kubernetes/kubernetes/issues/109905
thanks,

Reply:
Because containerd does not implement the ContainerStatus CRI extension, so I used a trickier way . When calling the ContainerStatus interface, if the verbose parameter is passed true, the spec information of the container will be returned, and the container resource can be obtained from the spec information.

Reply:
Because containerd does not implement the ContainerStatus CRI extension, so I used a trickier way . When calling the ContainerStatus interface, if the verbose parameter is passed true, the spec information of the container will be returned, and the container resource can be obtained from the spec information.

Reply:
Thank you very much for your approval, when your pr is merged, I will create a new pr based on your pr.

Because containerd does not implement the ContainerStatus CRI extension, so I used a trickier way . When calling the ContainerStatus interface, if the verbose parameter is passed true, the spec information of the container will be returned, and the container resource can be obtained from the spec information.The following is the sample code:

```
func parseContainerResource(containerID string, containerInfoMap map[string]string) (*runtimeapi.LinuxContainerResources, error) {
	infoStr, ok := containerInfoMap["info"]
	if !ok {
		return nil, nil
	}

	containerInfo := &ContainerInfo{}
	if err := json.Unmarshal([]byte(infoStr), containerInfo); err != nil {
		klog.Errorf("[RemoteRuntimeService] ContainerStatus Response Unmarshal containerInfo (containerID=%+v, info=%+v) err %+v ", containerID, infoStr, err)
		return nil, err
	}

	if containerInfo.RuntimeSpec == nil || containerInfo.RuntimeSpec.Linux == nil || containerInfo.RuntimeSpec.Linux.Resources == nil {
		return nil, nil
	}
	resources := &runtimeapi.LinuxContainerResources{}
	cpu := containerInfo.RuntimeSpec.Linux.Resources.CPU
	memory := containerInfo.RuntimeSpec.Linux.Resources.Memory

	if cpu != nil {
		if cpu.Period != nil {
			resources.CpuPeriod = int64(*cpu.Period)
		}
		if cpu.Quota != nil {
			resources.CpuQuota = *cpu.Quota
		}
		if cpu.Shares != nil {
			resources.CpuShares = int64(*cpu.Shares)
		}
	}

	if memory != nil {
		if memory.Limit != nil {
			resources.MemoryLimitInBytes = *memory.Limit
		}
	}

	return resources, nil
}
```

Then I set the resource to resp.Status.Resources,I found this code to work fine

Then I ran the e2e test, using the resizeToOriginalPatchString parameter, and still reported the same error, the error is as follows
```
[sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling]
/home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/framework.go:23
  Guaranteed QoS pod, one container - increase CPU & memory [It]
  /home/zuoshaoyi/go/src/k8s.io/kubernetes.vpa/test/e2e/node/pod_resize.go:1151

  May 11 10:57:30.217: failed to verify container cgroup values equals expected values
  Unexpected error:
      <*errors.errorString | 0xc000bec300>: {
          s: "timed out waiting for container cgroup values to match expected",
      }
      timed out waiting for container cgroup values to match expected
  occurred
```

I think the possible reason is that ```startContainer``` function calls ```newContainerAnnotations``` function to set ```io.kubernetes.container.hash```, and the hash value will not be updated when the pod resource is updated subsequently. So when resize resource A -> resource B success, and then resize resource B -> resource A , the hash value will not change,The following code will not execute.I think this is the reason why e2e execution fails

```
if _, _, changed := containerChanged(&container, containerStatus); changed {
	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) &&
		kubecontainer.HashContainerWithoutResources(&container) == containerStatus.HashWithoutResources {
		// Only the Resources field has changed. Check for in-place pod resize action.
		if keep := m.computePodResizeAction(pod, idx, containerStatus, &changes); keep {
			keepCount++
		}
		continue
	} else {
		message = fmt.Sprintf("Container %s definition changed", container.Name)
		// Restart regardless of the restart policy because the container
		// spec changed.
		restart = true
	}
}
```

Reply:
@zuoshaoyi Please excuse my slow responses.. this PR has been just a side project for me for quite some time with other priorities often preempting it. But I tried out your test and you are correct. You caught a very good corner case here. Perhaps we should limit 'restart on hash change' check only to exclude cases where it is 'Resources only' change. (This is a temporary change to handle feature gate toggle in alpha & beta without affecting all containers on the node). I think we can call computePodResizeAction at the bottom of the chain when there are no other reasons requiring container restart. I'll take a closer look when I get the chance.

As for extending the E2E test, I think we could use two things:
1. Extend the current tests to patch back to original after successful resize.
2. Patch pod forward and patch it back to original without waiting for resize - perhaps with random delay in between - with the goal of verifying that actual state eventually converges to the desired regardless of how far resize progressed. We can reuse the same bank of test-cases.

In general, I we should avoid making the tests specific to cgroups as much as possible (we want this to work for other platforms as well - so use standard APIs). But at this point given our unique chicken-and-egg situation, it is ok to assume containerd and check cgroups as that is the only verification we can get to truly know that resize was actuated until runtime support gets added (by beta timeframe). But I expect these tests will evolve by feature GA.

Reply:
@zuoshaoyi Please review the change https://github.com/kubernetes/kubernetes/pull/102884/commits/962cb2aabff460d19bc077d14c44e6864bf07ea9 which addresses the issue you found. You can still create e2e test PR but needs to be activated after beta ( or rather after @ruiwen-zhao containerd [change](https://github.com/containerd/containerd/pull/6517) is merged and in-play in k8s) 

With his change, you should be able to see the test pass but needs the custom containerd until that PR gets in.